pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -199,32 +223,238 @@ pub mod pallet {
+			let burnt = T::Fungibles::burn_from(
+				asset_id.clone(),
+				&who,
+				Bounded::max_value(),
+				Precision::BestEffort,
+				Fortitude::Force,
+			)?;
+
+			// no funds available to be burnt on account; nothing to do here
+			ensure!(!burnt.is_zero(), Error::<T>::ZeroAmount); // TODO: fail or return?
+
+			let sum_of_issuances = pool_details
+				.bonded_currencies
+				.into_iter()
+				.fold(FungiblesBalanceOf::<T>::zero(), |sum, id| {
+					sum.saturating_add(T::Fungibles::total_issuance(id))
+				});
+
+			let amount = burnt
+				.checked_mul(&total_collateral_issuance)
+				.ok_or(ArithmeticError::Overflow)? // TODO: do we need a fallback if this fails?
+				.checked_div(&sum_of_issuances)
+				.ok_or(Error::<T>::ZeroAmount)?; // should be impossible - how would we be able to burn funds if the sum of total supplies is 0?
+
+			if amount.is_zero()
+				|| T::CollateralCurrency::can_deposit(T::CollateralAssetId::get(), &who, amount, Provenance::Extant)
+					.into_result()
+					.is_err()
+			{
+				// funds are burnt but the collateral received is not sufficient to be deposited to the account
+				// this is tolerated as otherwise we could have edge cases where it's impossible to refund at least some accounts
+				return Ok(());
+			}
+
+			let transferred = T::CollateralCurrency::transfer(
+				T::CollateralAssetId::get(),
+				&pool_account,
+				&who,
+				amount,
+				Preservation::Expendable,
+			)?; // TODO: check edge cases around existential deposit
+
+			// if collateral or total supply drops to zero, refunding is complete -> emit event
+			if sum_of_issuances <= burnt || total_collateral_issuance <= transferred {
+				Self::deposit_event(Event::RefundComplete(pool_id));
+			}
+
+			Ok(())
 		}
 
 		#[pallet::call_index(9)]
 		#[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
-		pub fn finish_destroy(_origin: OriginFor<T>) -> DispatchResult {
-			todo!()
+		pub fn start_destroy(origin: OriginFor<T>, pool_id: T::PoolId) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			Self::do_start_destroy_pool(pool_id, Some(&who), Fortitude::Polite)
+		}
+
+		#[pallet::call_index(10)]
+		#[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
+		pub fn force_start_destroy(origin: OriginFor<T>, pool_id: T::PoolId) -> DispatchResult {
+			ensure_root(origin)?;
+
+			Self::do_start_destroy_pool(pool_id, None, Fortitude::Force)
+		}
+
+		#[pallet::call_index(11)]
+		#[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
+		pub fn finish_destroy(origin: OriginFor<T>, pool_id: T::PoolId) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			let pool_details = Pools::<T>::get(pool_id.clone()).ok_or(Error::<T>::PoolUnknown)?;
+
+			ensure!(pool_details.state.is_destroying(), Error::<T>::LivePool);
+
+			for asset_id in pool_details.bonded_currencies {
```

Comment:
Once again, this is also hard to benchmark. Can't we simply fail to finish the destroy if there's any currencies left, i.e., the counter I proposed above is grater than `0`?

Reply:
The only way to avoid looping is to call this or an additional extrinsic once per currency, which would we much worse UI than just having another parameter on the extrinsic to which you need to pass the number of currencies for weight calculation. So I'd rather do the latter

Reply:
Works for me.

Reply:
93f29503e5ccbeee4157ac08459d0a0eb92bb98e